### ISSUE TITLE
About hash string header bytes

### ISSUE THREAD

**Issue by lyrachord** [trendmicro/tlsh#37]
Hi,  the hash has 6 half-byte chars: checksum, Lvalue and Quartile, 3 bytes.
I have 2 questions:
1, why checksum is a byte array?  it's just a single byte.
2, why do swapping for the head bytes? historical reasons?

Regargs

**Comment by lyrachord** [comment]
in tlsh_impl.h, more versions of TLSH_CHECKSUM_LEN found, so Q1:) is solved

**Comment by jonjoliver** [comment]
Could you explain what you mean by "swapping for the head bytes" ?

Cheers
jono

**Comment by lyrachord** [comment]
Thanks jono

> swapping for the head bytes

I mean, swap_byte invocation in hash function 

f.e. a digest with checksum[1], Lvalue, Q, which head hexes are 3 bytes: 
>C1C2 L1L2 Q1Q2 

function hash() will return a string which starts with 
>C2C1L2L1Q2Q1.

there is a swapping-first, why? 

Regards

**Comment by jonjoliver** [comment]
I think the swap byte invocation is required because different OS store the bytes in a different order
And the different calls allow the programs to get access to the header in a consistent logical view independent of the OS implementation of memory

**Comment by lyrachord** [comment]
Thanks.
It seems it's a old history. You know, new CPUs care no a single byte, just byte order, even more MCUs.
I guess, to replace the swapping way with current normal byte hex string representation and to keep the old hash in history function is not bad idea?

Regards

**Comment by jonjoliver** [comment]
Hi lyrachord
I don't think I will change the header code - I think it works in some older implementations
And as far as I am aware - the code as it is now - is consistent across OS - and is not breaking.
Cheers
jono

**Comment by lyrachord** [comment]
Hi jono 
Breaking or not is a problem. You're right.
Let's in other words, Keeping the swapping and add a new version without swapping hash2 function?

BTW, you know the TLSH-java project deleted, I'm planning to contribute another java version. But there is naming convention inconsistent problem between java and c/c++, your advice?

Regards

**Comment by mrpolyonymous** [comment]
Hi @lyrachord 
I just contributed a Java port to the TLSH project here, it's in https://github.com/trendmicro/tlsh/tree/master/java

It will look very familiar if you've looked at the TLSH C++ code, and for good or bad I resisted making the variable and method names more Java-like to keep the similarity to the C++ code. 

The library isn't yet on Maven central or any other JAR hosting site, that's my next task. If you're okay building locally then running `gradlew build` will make the JAR file for you.
